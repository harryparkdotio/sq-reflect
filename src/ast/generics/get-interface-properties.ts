import * as ts from 'typescript';

import { Transform } from '..';

export const getGenericInterfacePropertiesFromTypeParameters = (
  attributeName: string,
  typeParameters: ts.TypeParameterDeclaration[]
): [ts.TypeParameterDeclaration[], ts.TypeReferenceNode[]] => {
  const interfaceTypeParameters: ts.TypeParameterDeclaration[] = [];
  const propertyTypeTypeArguments: ts.TypeReferenceNode[] = [];

  typeParameters.forEach((tp, idx) => {
    const interfaceTypeParameterIdentifier = ts.createIdentifier(
      Transform.interfaceTypeParameterName(
        typeParameters.length > 1
          ? `${attributeName}_${idx.toString()}_type`
          : `${attributeName}_type`
      )
    );

    interfaceTypeParameters.push(
      ts.createTypeParameterDeclaration(
        interfaceTypeParameterIdentifier,
        undefined,
        tp.default
      )
    );

    propertyTypeTypeArguments.push(
      ts.createTypeReferenceNode(interfaceTypeParameterIdentifier, undefined)
    );
  });

  return [interfaceTypeParameters, propertyTypeTypeArguments];
};
