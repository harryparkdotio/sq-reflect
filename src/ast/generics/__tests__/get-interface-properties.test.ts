import * as ts from 'typescript';
import { getGenericInterfacePropertiesFromTypeParameters } from '../get-interface-properties';

describe('getGenericInterfacePropertiesFromTypeParameters', () => {
  test('should', () => {
    expect(
      getGenericInterfacePropertiesFromTypeParameters('abcd', [
        ts.createTypeParameterDeclaration(
          'aaa',
          undefined,
          ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword)
        ),
      ])
    ).toMatchSnapshot();
  });

  test('should', () => {
    expect(
      getGenericInterfacePropertiesFromTypeParameters('abcd', [
        ts.createTypeParameterDeclaration(
          'abc',
          undefined,
          ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword)
        ),
        ts.createTypeParameterDeclaration(
          'def',
          undefined,
          ts.createUnionTypeNode([
            ts.createKeywordTypeNode(ts.SyntaxKind.StringKeyword),
            ts.createKeywordTypeNode(ts.SyntaxKind.ObjectKeyword),
          ])
        ),
      ])
    ).toMatchSnapshot();
  });
});
