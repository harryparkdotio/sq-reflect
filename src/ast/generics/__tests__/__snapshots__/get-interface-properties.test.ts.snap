// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getGenericInterfacePropertiesFromTypeParameters should 1`] = `
Array [
  Array [
    AbcdType = any,
  ],
  Array [
    AbcdType,
  ],
]
`;

exports[`getGenericInterfacePropertiesFromTypeParameters should 2`] = `
Array [
  Array [
    Abcd_0Type = any,
    Abcd_1Type = string | object,
  ],
  Array [
    Abcd_0Type,
    Abcd_1Type,
  ],
]
`;
